File: 133__Clone_Graph.py
# Definition for a undirected graph node
# class UndirectedGraphNode:
#     def __init__(self, x):
#         self.label = x
#         self.neighbors = []

class Solution:
    # @param node, a undirected graph node
    # @return a undirected graph node
    def cloneGraph(self, node):

        if node is None:
            return None

        newNodes = {}

        def dfs(node):
            if node.label in newNodes:
                return newNodes[node.label]
            else:
                temp_node = UndirectedGraphNode(node.label)
                newNodes[node.label] = temp_node
                for n in node.neighbors:
                    temp_node.neighbors.append(dfs(n))
                return temp_node

        return dfs(node)

File: Java/133__Clone_Graph.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;

    public Node() {}

    public Node(int _val,List<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/
class Solution {

    HashMap<Integer, Node> map = new HashMap<>();

    public Node cloneGraph(Node node) {
        help(node);
        return map.get(node.val);
    }

    private void help(Node node) {
        if (map.containsKey(node.val))
            return;
        Node temp = new Node(node.val);
        temp.neighbors = new ArrayList<Node>();
        map.put(temp.val, temp);
        for (Node origin_neighbor : node.neighbors) {
            if (map.containsKey(origin_neighbor.val))
                temp.neighbors.add(map.get(origin_neighbor.val));
            else {
                help(origin_neighbor);
                temp.neighbors.add(map.get(origin_neighbor.val));
            }
        }
    }
}
